WIP: Fixing broken handling of XpressNet turnouts - OFF messages.

svatoun opened the pull request:
This PR is still a work-in-progress. 
I am reorganizing code that was originally developed to completely fix the flaky handling of Accessory Feedback messages, but realized that handling of OFF messages may have value on it own, before that larger part will be completed.

I am trying to reorganize the source state into separate commits to make the history cleaner. First commit should just add diagnostics:
1. each message instance gets its `msg-id` -- unique hex string. It will be possible to search in the log to see how that particular message instance is processed. Particularly useful for general OK replies (all same content) which are scheduled-later.
2. I made a simple chain between command-replies:
    - if a command is issued **during** processing of a reply, it will connect back to that reply as its trigger.
    - a reply that comes **during* a message-send step in the transmission thread will connect back to that sent message as it's likely to repond to that message.
3. message's `unsolicited` flag is printed; helps to identify replies marked as unsolicited/broadcasts by the command station vs. messages just unrecognized by JMRI.

This information is printed on DEBUG level, and allows to quickly scan for trigger-action in the system log.

---
The other commit enhances tests. Simple command station simulators are created for **LZV100** (both USB and nonUSB connector), **DR5000**, **Nano-X**. Because of that, some methods in 
`XNetSimulatorAdapter` have to be published / refactored.

I have added tests, that check the behaviour of `XNetTrafficController` and/or `XNetTurnout`. I hope no leftovers from the fix developed are present - if so, sorry for that; the point was to show on a testsuite that the current behaviour is somwhat screwed up.

In particular:
* a test, that sends both normal and high-priority messages fails, since high-priorities are actually do not preempt normal messages.
* a test, that checks that first OFF message is delayed fails; (some) OFF message appears immediately after the reply

There are number of tests that check that 2 OFF messages are sent. They all fail as the current code sends 1-3 OFFs per reply because of coincidences, like invoking the same method recursively.

svatoun commented:
There will be another commit that fixes priorities, fixes OFF message to exactly one per reply and at least 2 per command (exactly 2, if just one feedback is received).

pabender commented on java/src/jmri/jmrix/lenz/xnetsimulator/XNetSimulatorAdapter.java:
```diff
@@ -483,18 +493,41 @@ private XNetReply accInfoReply(XNetMessage m) {
+     * @param msg    the output byte stream
+     * @param offset the first byte not yet used
+     */
+    protected void addTrailerToOutput(byte[] msg, int offset, XNetReply m) {
```
XpressNet messages should not have a trailer of any kind.  All XpressNet messages are binary, so the 0x0d You are adding to non-binary messages will never be added.

svatoun replied:
Thanks, a stupid merge error when pulling (I though) minimal changes to the PR. Luckily all XpressNets are binary, but must remove this.

pabender commented on java/test/jmri/jmrix/lenz/XNetTrafficControllerTest.java:
```diff
@@ -11,24 +27,321 @@
+    final Semaphore messageQueuePermits = new Semaphore(0);
+    volatile boolean timeoutOccured;
+    
+    XNetTurnoutManager xnetManager;
```
The tests that use the turnout manager in this file are in the wrong location.  So that they are identifiable as testing something related to the turnout ( they are not unit tests for the traffic controller ) They should either be in the XNetTurnoutManagerTest or the XNetTurnoutTest.  Alternatively, you can add a new test class called XNetTurnoutIT which will identify the tests as integration tests.

pabender commented:
Be careful when fixing the tests here.

Many of the XpressNet tests that are failing are doing so because of the way tests are inherited in the XpressNet code.  Some of these failures are due to tests running for some of the special case instances in the XpressNet code that was created for systems that don’t behave like an LZ100 or LZV100 connected to an LI100F.  

This is one of the reasons why I suggest moving tests to a different file, especially the ones you put in the XNetPacketizerTret class.

svatoun commented:
> This is one of the reasons why I suggest moving tests to a different file, especially the ones you put in the XNetPacketizerTest class.

Yes: I've realized that late - when trying to finalize the patch the work, tryint to run the whole testsuite. The other tests **could** be fixed eventually, since the only issue here was that  `setUp` and `tearDown` from the superclass is not called properly -- when I changed that, those test subclasses started to work.

So what was the intent when subclassing testcsaes: assure that even the subclass will satisfy all the states checked by the superclass' test - i.e. to check possible ill side effects of overriden methods ? If so, then I should make sure that the new tests actually run even with the subclasses somehow - except e.g. verifying transitions through states, which are disabled for `HornbyElite` etc.

svatoun commented:
I made a race condition in one of the tests: the intention was to have two normal messages in the transmit queue, and, between them, insert some high priority messages + normal messages. Then observe that the existing message that was not fetched is preemted, and that the newly inserted normal message will be sent last (after all high-prio and the current normals).

However :) the test main thread did not wait until the 1st normal message was taken ... 
... and did bad assert as well. So will correct and comment in code.

pabender commented on java/test/jmri/jmrix/lenz/LenzXNetTurnoutCommandTest.java:
```diff
@@ -0,0 +1,365 @@
+ * Checks Lenz clone's specific behaviour for Turnout commands.
+ * Assures that the Turnout will send an appropriate number of Accessory Request
+ * Output OFF commands, delayed properly and that it won't ping indefinitely.
+ */
```
Make sure you include author and copyright information here.

svatoun replied:
Isn't there a policy to assign Copyright to JMRI representative for easier legal processes ?

pabender replied:
We still mark the code with the author and copyright date.

pabender commented:
> > This is one of the reasons why I suggest moving tests to a different file, especially the ones you put in the XNetPacketizerTest class.
> 
> Yes: I've realized that late - when trying to finalize the patch the work, tryint to run the whole testsuite. The other tests **could** be fixed eventually, since the only issue here was that `setUp` and `tearDown` from the superclass is not called properly -- when I changed that, those test subclasses started to work.
> 

It is intentional that the setup is different for the subclasses.  If the setup were the same, then the tests would be testing the superclass objects rather than derived objects.

> So what was the intent when subclassing testcsaes: assure that even the subclass will satisfy all the states checked by the superclass' test - i.e. to check possible ill side effects of overriden methods ? If so, then I should make sure that the new tests actually run even with the subclasses somehow - except e.g. verifying transitions through states, which are disabled for `HornbyElite` etc.

The test subclassing generally verifies that the same contract is followed for each object subclass.  The object subclasses don’t follow the same behavior necessarily, so tests of specific behavior don’t always apply, or don’t apply in the same way, so the subclasses may model the actions that occur after an initial trigger differently.

svatoun commented:
I moved off XNetTurnout off-sequence specific tests, but I still don't know whether that's a good idea: except i.e. HornbyElite, it's probably desirable to verify that e.g. Z21 turnouts maintain the asserted invariants. If so, their tests could derive from the new Lenz-specific baseclass rather than `XNetTurnoutTest`.

svatoun commented:
@bobjacobsen, @pabender : I have noticed a different number of failed tests between AppVeyor and Headless tests. It seems that the `XNetTrafficControllerIT`` integration test is not run in the `headless` suite. Is that OK ?

pabender commented:
> I moved off XNetTurnout off-sequence specific tests, but I still don't know whether that's a good idea: except i.e. HornbyElite, it's probably desirable to verify that e.g. Z21 turnouts maintain the asserted invariants. If so, their tests could derive from the new Lenz-specific baseclass rather than `XNetTurnoutTest`.

The Z21 case is, like the HornbyElite case, a situation where the manufacturer did something different than what Lenz did.  The response sequence is different.

pabender commented:
> @bobjacobsen, @pabender : I have noticed a different number of failed tests between AppVeyor and Headless tests. It seems that the `XNetTrafficControllerIT` integration test is not run in the `headless` suite. Is that OK ?

Which headless?  "ant headlesstest" running locally or the headless tests that runs on Travis?

This is probably a configuration oversight on my part.  I will fix.

pabender commented:
> > @bobjacobsen, @pabender : I have noticed a different number of failed tests between AppVeyor and Headless tests. It seems that the `XNetTrafficControllerIT` integration test is not run in the `headless` suite. Is that OK ?
> 
> Which headless? "ant headlesstest" running locally or the headless tests that runs on Travis?
> 
> This is probably a configuration oversight on my part. I will fix.

I think i see what is going on.

In maven, the integration tests only run if the unit tests pass.  The unit tests are failing, so the integration tests never start.

In ant, we are running all the tests as part of a single test suite.  We could separate out the integration tests there as well.

So it is running as intended.

svatoun commented:
@pabender ah, thanks !

svatoun commented:
I think the 7d2b620 provides a baseline for analysis / comparison with the fixes. The tests can be debugged to inspect closely why the testcases break. I still need to look on several AppVeyor results to ensure that I got the timing right - coordinating threads is tricky and if not done well, will break on another build machine.

svatoun commented:
Well :) one case escaped: for the case of consecutive turnouts, where XT11 receives a feedback reply triggered (on DR5000) by subsequent command on XT12. At that time, it is indistinguishable from the multiple-feedback to the same turnout (yet).
Two options here: leave it as it is, or limit the number of send off-s strictly to two, no matter how many feedback messages comes before first OK-while-OFFSENT.

The case should be also suppressed later (different PR), if I succeed to build a heuristic that will (hopefully) fix the Nano-X issues for good.

svatoun commented:
A few comments:

#### Priority flag
I've introduced it on `AbstractMRMessage` + handling in `AbstractMRTrafficController`, as also `DCCppTrafficController` derives from that class, and has copied the entire priority stuff from XNet.
The reason for having the flag on the message is diagnostics; prints out as `priority` in logs.

#### Reply consumed
The `XNetReply` message traverses through (potentially) many listeners, finally reaches its target (if applicable) as a sole context object; I introduced a consumed flag similar to other event systems. This flag(s) is checked / used to produce at most one OFF message during this reply processing. There's special handling for feedbacks.

#### extra OFFSENT state
On the mailing list it was said that there must be 2 messages - this can be encoded in state transitions (instead of some counter). 

#### sendOffMessage trampoline
`protected void sendOffMessage()` has been there for several releases. The (deprecated) trampoline that calls the real implementation maintains compatible interface to possible subclasses out there.

#### tracking 'last commanded'
During debugging, it turned out that when commading consecutive turnouts, the feedback in the pair may temporarily obscur the turnout's KnownState and CommandedState; the `sendOffMessage` sends one to three messages for the wrong output - logs provided on mailing list, I have also logs form LZV100 operating the same way. Given that (because of queues) only one switch command can be active at a time, it's possible to send OFF message for just the output operated by JMRI.

Original course of events can be still observed or debugged for comparison, if certain methods are changed to stubs:
* If  `AbstractMRReply::markConsumed` and `isConsumed` return  `false`, the original number of OFF messages will appear.
* If `XNetTurnout::sendQueuedMessage()` does not set `lastLayoutCommanded`, the OFF messages will be forrmed using the original way
* If conditional branch at `XNetTurnout:641` is commented out, each feedback will produce an OFF message, so there will be more than 2 of them, eventually.

#### Minor stuff
I've optimized the output message queue in `XNetTurnout`. It was originally highly concurrent impl, but was required to be operated only under a shared lock with `internalState`, so that sophisticated implementation was practically useless.
In addition, although the queue was not size-limited which would block on `offer`, for some reason blocking method variants were used instead of non-blocking: unncessary try-catch handling etc.

svatoun commented:
seems that I should have run more tests locally. Setting as draft.

svatoun commented:
Need to add some reliable way how to wait until all the startup commands settle in the integration test. Last failure was a regular startup command station status query that came late.

pabender commented on java/src/jmri/jmrix/lenz/XNetReply.java:
```diff
@@ -349,6 +369,110 @@ public int getFeedbackEncoderMsgAddr() {
+        if (isFeedbackMessage()) {
+            return 1;
+        } else if (isFeedbackBroadcastMessage()) {
+            return (this.getElement(0) & 0x0F);
```
That isn’t right.  The second Nibble of the op code is the number of data bytes that follows.  For feedback messages ( directed or broadcast ) each feedback address occupies two bytes. ( an address byte and a data byte ) so the number of feedback values in the message is `(getElement(0)&0x0F)/2`

pabender commented on java/src/jmri/jmrix/lenz/XNetReply.java:
```diff
@@ -686,6 +810,12 @@ public boolean isUnsolicited() {
     }
 
+    /**
+     * Resets the unsolicited feedback flag. If the feedback was received
```
Why’ll feedback messages may be the primary user of this, this doesn’t just apply to feedback messages.

pabender commented on java/src/jmri/jmrix/lenz/hornbyelite/EliteXNetTurnout.java:
```diff
@@ -22,7 +22,7 @@ public EliteXNetTurnout(String prefix, int pNumber, XNetTrafficController tc) {
      */
     @Override
-    protected synchronized void sendOffMessage() {
+    protected synchronized void doSendOffMessage() {
```
Send is a verb.  The do is redundant.

svatoun replied:
pls. suggest a better naming; I did not come up with anything 'sane' and plain `sendOffMessage` was taken. I'm used to use this pattern, or e.g. `xxxProtected`, `performXXX`, but don't know what's the preference here.

pabender replied:
You renamed sendOffMessage as doSendOffMessage.  I would name it back.

svatoun replied:
@pabender looking backwards after few days, I've outsmarted myself with the backward-compatibility-trampoline stuff - after the refactorings it does not bring any value.  Will revert back to just `sendOffMessage` as you suggest.

pabender commented on java/src/jmri/jmrix/lenz/XNetTrafficController.java:
```diff
@@ -140,45 +138,17 @@ public void forwardReply(AbstractMRListener client, AbstractMRReply m) {
-            }
-        } catch (java.lang.InterruptedException ie) {
-            log.error("Interrupted while removing High Priority Message Listener from Queue");
-        }
```
If pollMessage and pollReplyHandler are now returning null, check and see if they can be removed.  I think there is a default implementation in AbstractMRTrafficController

svatoun replied:
OTOH, there are abstracts. Can be changed to no-op methods.

pabender commented:
> #### Reply consumed
> The `XNetReply` message traverses through (potentially) many listeners, finally reaches its target (if applicable) as a sole context object; I introduced a consumed flag similar to other event systems. This flag(s) is checked / used to produce at most one OFF message during this reply processing. There's special handling for feedbacks.
> 

I am not sure I like the consumed terminology here.  When something is consumed, it generally is considered unavailable for further processing, but that isn’t always the case in JMRI ( or most other event driven systems I am familiar with ).

Each of the XpressNet listeners Listens to the conversation and performs actions they see as necessary.  This means multiple listeners can process, or consume, the information.  Which one of those listeners marks the message as consumed?

One obvious listener that uses the replies to perform actions is the XpressNet monitor, but users are allowed to written listeners as scripts and perform their own actions on the message.  Users could certainly see the methods related to the consumed flag and mark them as consumed.

svatoun commented:
> I am not sure I like the consumed terminology here. When something is consumed, it generally is considered unavailable for further processing, but that isn’t always the case in JMRI ( or most other event driven systems I am familiar with ).

Yes, that's why the 'consumed' should be rather a flag. I.e. in Swing, an event is also marked as consumed, but may be seen afterwards.

Question: is the objection primarily a terminology one - or conceptua / accessibility one:

ad terminology: can be better specified in documentation: i.e. "those bits + those functions are used by XpressNet subsystem itself, you should not use those values, unless you want to step into the protocol processing".

ad exposing the concept: 
I wanted (primarily) to make sure XpressNet clients can either guard themselves (using different bits) or pre-filter things before they reach the target. The primary target was obviously `XNetTurnout`. 

Also the "heuristics" I was talking about (in development, immature) selectively ignores some feedbacks if the accessory command is known to be sent (already queued). I planned to do so by communicating with XNetTurnout implementation using the "context" object (`XNetReply`). Don't worry ;) part of the heuristic should be a conservative query to the layout, if a potential race condition is detected.

But both of the clients are 'intrinsic' to XpressNet subsystem, so this support can be sealed if seen necessary. But since similar complexities as in XNetTurnout may happen elsewhere, I just thought it could be useful in general.

If documentation warning is not sufficient, then:
- all the consumed-related stuff might be made package-private, so it is only accessible to the xpressnet core. Could be problematic in specializations (Z21 et all), if it turns out that they need to override for specific cases.
- the methods can be made `protected`: only an implementator of `TrafficController`, who creates the replies can add accessor and its package (depending on accessor modifiers) can then step in.
- factor this out to a subclass of XNetReply, which is marked as "dangerous" (must be still public for subclassing, IMHO). Or to a separate Context object that can be looked up for the Reply instance. Normal clients won't see those methods, but with an effort, they can use them.

pabender commented:
> > I am not sure I like the consumed terminology here. When something is consumed, it generally is considered unavailable for further processing, but that isn’t always the case in JMRI ( or most other event driven systems I am familiar with ).
> 
> Yes, that's why the 'consumed' should be rather a flag. I.e. in Swing, an event is also marked as consumed, but may be seen afterwards.
> 
> Question: is the objection primarily a terminology one - or conceptua / accessibility one:
> 

My objection is primarily to the terminology.  Overloading Common term can be problematic.

The secondary concern is about how this is actually used.

Paul

svatoun commented:
> > Question: is the objection primarily a terminology one - or conceptua / accessibility one:

Please let's take it from this one
> The secondary concern is about how this is actually used.

The core of the change is to use a collaborative context during Reply processing to make sure that some actions (specifically: schedule or send "off" message) happen at most once per reply packet. Are there issues with the concept itself ?

pabender commented:
> The core of the change is to use a collaborative context during Reply processing to make sure that some actions (specifically: schedule or send "off" message) happen at most once per reply packet. Are there issues with the concept itself ?

I don't have an objection to the concept, but, personally, I am more concerned that the off messages arrive at the command station than the exact number we send.

svatoun commented:
So - what's the 'safe number' of packets or the strategy ? 

So far the code was wrtitten rather "coincidentally", so multiple paths may or may not trigger the OFF packet which may or may not be for the output (direction) previously commanded ON.

Communication with Lenz LZV100/LI-USB actually shown that (in some cases, turnout interference) out of 4 OFF packets sent, 3 used an invalid output number - which may or may not have the desired effects, depending on the decoder (and command station of course) implementation.

With this change in place, the number of OFFs can be set or controlled: if some command station needs more OFFs, the turnout can play ping-pong, sending another OFF to the previous one's OK reply as long as it is necessary.

pabender commented:
> So - what's the 'safe number' of packets or the strategy ?

We send off messages until we get an OK message in response to an off message.

> Communication with Lenz LZV100/LI-USB actually shown that (in some cases, turnout interference) out of 4 OFF packets sent, 3 used an invalid output number - which may or may not have the desired effects, depending on the decoder (and command station of course) implementation.
> 

For Lenz hardware, the output number doesn’t actually matter for the off messages.  Remember that Lenz LZ100/LZV100/LZV200 command stations stop repeating commands to the accessory decoder address with the off message.  Other lenz hardware completely ignores the off messages.

This could matter for other hardware, so that should be fixed.

What I am not concerned about at all is the 3 messages vs 4 messages.
